ci/infra/testrunner/platforms/terraform.py

```diff
@@ -2,143 +2,65 @@
-            if num_master > -1:
-                self.conf.master.count = num_master
-                logger.warning("   Masters:{} ".format(num_master))
-
-            if num_worker > -1:
-                self.conf.worker.count = num_worker
-                logger.warning("   Workers:{} ".format(num_worker))
-
+        exception = None
+        print("Init terraform")
         self._check_tf_deployed()
-        
+
         self.utils.setup_ssh()
 
-        init_cmd = "terraform init"
+        init_cmd = "init"
         if self.conf.terraform.plugin_dir:
-            logger.info("Installing plugins from {}".format(self.conf.terraform.plugin_dir))
-            init_cmd = init_cmd+" -plugin-dir="+self.conf.terraform.plugin_dir
-        self._runshellcommandterraform(init_cmd)
+            logger.info(f"Installing plugins from {self.conf.terraform.plugin_dir}")
+            init_cmd += f" -plugin-dir={self.conf.terraform.plugin_dir}"
+        self._run_terraform_command(init_cmd)
 
-        self._runshellcommandterraform("terraform version")
+        self._run_terraform_command("version")
         self._generate_tfvars_file()
-        plan_cmd = ("{env_setup};"
-                    " terraform plan "
-                    " -out {workspace}/tfout".format(
-                        env_setup=self._env_setup_cmd(),
-                        workspace=self.conf.workspace))
-        apply_cmd = ("{env_setup};"
-                     "terraform apply -auto-approve {workspace}/tfout".format(
-                        env_setup=self._env_setup_cmd(),
-                        workspace=self.conf.workspace))
-
-        # TODO: define the number of retries as a configuration parameter
-        for retry in range(1, 5):
-            self._runshellcommandterraform(plan_cmd)
-            try:
-                self._runshellcommandterraform(apply_cmd)
-                break
-
-            except Exception as ex :
-                logger.warning("Failed terraform apply attempt {}/5".format(retry))
-                if retry == 4:
-                    raise Exception("Failed Openstack Terraform deployment") from ex
-            finally:
-                self._fetch_terraform_output()
-
-    @staticmethod
-    def _create_node_log_dir(ip_address, node_type, log_dir_path):
-        node_log_dir_path = os.path.join(log_dir_path, f"{node_type}_{ip_address.replace('.', '_')}")
-
-        if not os.path.isdir(node_log_dir_path):
-            os.mkdir(node_log_dir_path)
-            logger.info(f"Created log dir {node_log_dir_path}")
-
-        return node_log_dir_path
+        plan_cmd = f"plan -out {self.tfout_path}"
+        apply_cmd = f"apply -auto-approve {self.tfout_path}"
+
+        self._run_terraform_command(plan_cmd)
+
+        try:
+            self._run_terraform_command(apply_cmd)
+        except Exception as ex:
+            exception = ex
+        finally:
+            self._fetch_terraform_output()
+            if exception:
+                raise exception
```

Comment:
The `finally` will re-raise an exception generated inside the `except` block, so you could put a bare`raise` in the except and save a couple lines.  But it will also re-raise any unhandled exception.  So, since all exceptions are ultimately being raised, this could be simplified as `try/finally`, leaving the except and both raises out.  Either way, call that optional cleanup for later. :)

Reply:
Oh haha yeah 😂